[Bug] V compiler not detecting type error for os.read_file "?string" type.

**V version:**
```cmd
> v -v
V 0.1.29 35cbca9
```
**OS:** Windows 10 2004

### What did you do?
```v
// ./hello.v:

import os

fn main() {
    print("./rand: " + os.read_file("./rand"))
}
```
```cmd
> v run "d:\projects\v\hello\hello.v"
```

### What did you expect to see?
An error indicating that I was adding incompatible types(`string` and `?string`).
 
### What did you see instead?
```cmd
==================
C:\Users\aqilc\AppData\Local\Temp\v\hello.tmp.c: In function 'main__main':
C:\Users\aqilc\AppData\Local\Temp\v\hello.tmp.c:10041:49: error: expected expression before 'Option_string'
  print(string_add(tos_lit("hewwo, heres rand"), Option_string _t191 = os__read_file(tos_lit("./rand"));
                                                 ^~~~~~~~~~~~~
C:\Users\aqilc\AppData\Local\Temp\v\hello.tmp.c: In function '_vinit':
C:\Users\aqilc\AppData\Local\Temp\v\hello.tmp.c:10070:38: warning: integer constant is so large that it is unsigned
  _const_math__bits__max_u64 = ((u64)(18446744073709551615));
                                      ^~~~~~~~~~~~~~~~~~~~
C:\Users\aqilc\AppData\Local\Temp\v\hello.tmp.c:10097:409: warning: integer constant is so large that it is unsigned
   ((u64)(1)), ((u64)(10)), ((u64)(100)), ((u64)(1000)), ((u64)(10000)), ((u64)(100000)), ((u64)(1000000)), ((u64)(10000000)), ((u64)(100000000)), ((u64)(1000000000)), ((u64)(10000000000)), ((u64)(100000000000)), ((u64)(1000000000000)), ((u64)(10000000000000)), ((u64)(100000000000000)), ((u64)(1000000000000000)), ((u64)(10000000000000000)), ((u64)(100000000000000000)), ((u64)(1000000000000000000)), ((u64)(10000000000000000000))}));
                                                                                                                                                                                                                                                                                                                                                                                                                         ^~~~~~~~~~~~~~~~~~~~
C:\Users\aqilc\AppData\Local\Temp\v\hello.tmp.c: In function 'wmain':
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.
```

--- a/vlib/v/checker/check_types.v
+++ b/vlib/v/checker/check_types.v
@@ -182,6 +182,9 @@ pub fn (c &Checker) promote(left_type, right_type table.Type) table.Type {
 	}
 	if right_type.is_number() && left_type.is_number() {
 		return c.promote_num(left_type, right_type)
+	} else if left_type.derive(right_type) != left_type {
+		// incompatible flags
+		return table.void_type
 	} else {
 		return left_type // default to left if not automatic promotion possible
 	}

--- a/vlib/v/checker/check_types.v
+++ b/vlib/v/checker/check_types.v
@@ -182,8 +182,8 @@ pub fn (c &Checker) promote(left_type, right_type table.Type) table.Type {
 	}
 	if right_type.is_number() && left_type.is_number() {
 		return c.promote_num(left_type, right_type)
-	} else if left_type.derive(right_type) != left_type {
-		// incompatible flags
+	} else if left_type.has_flag(.optional) != right_type.has_flag(.optional) {
+		// incompatible
 		return table.void_type
 	} else {
 		return left_type // default to left if not automatic promotion possible

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -618,6 +618,10 @@ pub fn (mut c Checker) infix_expr(mut infix_expr ast.InfixExpr) table.Type {
 					left_name := c.table.type_to_str(left_type)
 					right_name := c.table.type_to_str(right_type)
 					c.error('mismatched types `$left_name` and `$right_name`', infix_expr.pos)
+				} else if promoted_type.has_flag(.optional) {
+					s := c.table.type_to_str(promoted_type)
+					c.error('`$infix_expr.op` cannot be used with `$s`',
+						infix_expr.pos)
 				} else if promoted_type.is_float() {
 					if infix_expr.op in [.mod, .xor, .amp, .pipe] {
 						side := if left_type == promoted_type { 'left' } else { 'right' }

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -620,8 +620,7 @@ pub fn (mut c Checker) infix_expr(mut infix_expr ast.InfixExpr) table.Type {
 					c.error('mismatched types `$left_name` and `$right_name`', infix_expr.pos)
 				} else if promoted_type.has_flag(.optional) {
 					s := c.table.type_to_str(promoted_type)
-					c.error('`$infix_expr.op` cannot be used with `$s`',
-						infix_expr.pos)
+					c.error('`$infix_expr.op` cannot be used with `$s`', infix_expr.pos)
 				} else if promoted_type.is_float() {
 					if infix_expr.op in [.mod, .xor, .amp, .pipe] {
 						side := if left_type == promoted_type { 'left' } else { 'right' }

--- a/vlib/v/checker/tests/infix_err.out
+++ b/vlib/v/checker/tests/infix_err.out
@@ -1,0 +1,39 @@
+vlib/v/checker/tests/infix_err.vv:7:8: error: mismatched types `string` and `?string`
+    5 |     return none
+    6 | }
+    7 | _ = '' + f()
+      |        ^
+    8 | _ = f() + ''
+    9 | _ = f() + f()
+vlib/v/checker/tests/infix_err.vv:8:9: error: mismatched types `?string` and `string`
+    6 | }
+    7 | _ = '' + f()
+    8 | _ = f() + ''
+      |         ^
+    9 | _ = f() + f()
+   10 |
+vlib/v/checker/tests/infix_err.vv:9:9: error: `+` cannot be used with `?string`
+    7 | _ = '' + f()
+    8 | _ = f() + ''
+    9 | _ = f() + f()
+      |         ^
+   10 |
+   11 | _ = 4 + g()
+vlib/v/checker/tests/infix_err.vv:11:7: error: `+` cannot be used with `?int`
+    9 | _ = f() + f()
+   10 |
+   11 | _ = 4 + g()
+      |       ^
+   12 | _ = int(0) + g() // FIXME not detected
+   13 | _ = g() + int(3)
+vlib/v/checker/tests/infix_err.vv:13:9: error: `+` cannot be used with `?int`
+   11 | _ = 4 + g()
+   12 | _ = int(0) + g() // FIXME not detected
+   13 | _ = g() + int(3)
+      |         ^
+   14 | _ = g() + 3
+vlib/v/checker/tests/infix_err.vv:14:9: error: `+` cannot be used with `?int`
+   12 | _ = int(0) + g() // FIXME not detected
+   13 | _ = g() + int(3)
+   14 | _ = g() + 3
+      |         ^

--- a/vlib/v/checker/tests/infix_err.vv
+++ b/vlib/v/checker/tests/infix_err.vv
@@ -1,0 +1,14 @@
+fn f() ?string {
+	return none
+}
+fn g() ?int {
+	return none
+}
+_ = '' + f()
+_ = f() + ''
+_ = f() + f()
+
+_ = 4 + g()
+_ = int(0) + g() // FIXME not detected
+_ = g() + int(3)
+_ = g() + 3

